wasi/src/http_client.rs

```diff
@@ -0,0 +1,200 @@
+use anyhow::{anyhow, Context};
+use bytes::{BufMut, Bytes, BytesMut};
+use http::header::{HeaderName, HeaderValue};
+#[cfg(feature = "std")]
+use std::ops::Deref;
+
+use crate::snapshots::preview_2::{default_outgoing_http, streams, types as http_types};
+
+pub struct DefaultClient {
+    options: Option<default_outgoing_http::RequestOptions>,
+}
+
+impl DefaultClient {
+    pub fn new(options: Option<default_outgoing_http::RequestOptions>) -> Self {
+        Self { options }
+    }
+
+    pub fn handle(&self, req: http::Request<Bytes>) -> anyhow::Result<http::Response<Bytes>> {
+        let req = Request::try_from(req)
+            .context("converting http request")?
+            .to_owned();
+
+        let res = default_outgoing_http::handle(req, self.options);
+        http_types::drop_outgoing_request(req);
+
+        let response =
+            http::Response::try_from(Response(res)).context("converting http response")?;
+        http_types::drop_incoming_response(res);
+
+        Ok(response)
+    }
+}
+
+pub struct Request(default_outgoing_http::OutgoingRequest);
+
+#[cfg(feature = "std")]
```

Comment:
If the above change is made, this `cfg` and the others below could also be removed.